### PR TITLE
[1.2.0] SQL Support [API-495] [API-1015] [API-864] [API-865] [API-867] [API-857] [API-858] [API-859] [API-860] [API-834] [API-835]

### PR DESCRIPTION
**Pull request by yuce**
This PR:
* Refactors `hazelcast.Client` into internal `client.Client` and other Client types which use it, `hazelcast.Client` and `driver.Connection`.
* Adds support for database/sql interface. See: https://pkg.go.dev/database/sql Currently Hazelcast connection is established eagerly on `sql.Open` (in contrast with some other SQL packages which creates connections on demand).
* Adds SQL Service which is used by the database/sql interface.
* Adds builtin / portable big int, decimal data/time serialization functions.
* Adds portable data types, decimal support
* Changes the connection logic to try all seed addresses.

### CODE REVIEW

File: internal/sql/driver/statement.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+func (s Statement) Close() error {
+	return nil
+}
+
+func (s Statement) NumInput() int {
+	return -1
+}
+
+func (s *Statement) Exec(args []driver.Value) (driver.Result, error) {
```
Comment (utku-caglayan) [inline internal/sql/driver/statement.go:44]: Can you also check [StmtExecContext](https://pkg.go.dev/database/sql/driver#StmtExecContext) and [StmtQueryContext](https://pkg.go.dev/database/sql/driver#StmtQueryContext)? 🙏
 
Because It looks like we have everything ready for context support 🚀

File: sql/example_test.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package sql_test
+
+import (
+	"context"
+	"database/sql"
+)
```
Comment (utku-caglayan) [inline sql/example_test.go:6]:
```suggestion
    _ "github.com/hazelcast/hazelcast-go-client/sql/driver"
)
```

File: sql/example_test.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		fmt.Println(result.Row.Columns[0])
+		// alternative: fmt.Println(result.Row.ColumnByIndex(0))
+		// alternative: fmt.Println(result.Row.ColumnByName("foo"))
+	}
+}
+
+*/
+
+func ExampleDriver_Open() {
+	db, err := sql.Open("hazelcast", "localhost;Cluster.Unisocket=true")
```
Comment (utku-caglayan) [inline sql/example_test.go:33]: It raises an error without port. Maybe we can use default port if not provided?

File: sql/example_test.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		panic(err)
+	}
+	ctx := context.Background()
+	conn, err := db.Conn(ctx)
+	if err != nil {
+		panic(err)
+	}
+	defer conn.Close()
+	/*
+		rows, err := conn.QueryContext(ctx, "SELECT name FROM emp WHERE age < ?", 30)
```
Comment (utku-caglayan) [inline sql/example_test.go:44]: I believe we can uncomment this

**Review by utku-caglayan — CHANGES_REQUESTED** [review]
Provided SQL API looks great! I have commented on some minor points

File: internal/sql/driver/driver_test.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+				Name: "dev",
+				Network: cluster.NetworkConfig{
+					Addresses:         []string{"10.20.30.40:5000", "11.21.31.41:5001"},
+					PortRange:         cluster.PortRange{Min: 5701, Max: 5703},
+					ConnectionTimeout: types.Duration(5 * time.Second),
+				},
+			},
+		},
+		{
+			DSN: "10.20.30.40:5000;ClusterName=my-cluster",
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply internal/sql/driver/driver_test.go:69]: We should also change example_test.go if we decide to change representation

File: examples/sql/driver/ping/main.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+func main() {
+	log.Println("Creating the database value.")
+	db, err := sql.Open("hazelcast", ";logger.level=info")
+	if err != nil {
+		panic(err)
+	}
+	defer db.Close()
+	log.Println("The database value was created. Note that the connection to Hazelcast is not established at this point.")
```
[the remark this answers is not in the document]
Comment (yuce) [reply examples/sql/driver/ping/main.go:33]: "Opening" the connection (creating/starting the client) is lazy, so `sql.Open` doesn't actually do that.

File: internal/sql/driver/sql_service.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+		return nil, ihzerrors.NewIOError("no connection found", nil)
+	}
+	qid := isql.NewQueryIDFromUUID(conn.MemberUUID())
+	req := codec.EncodeSqlExecuteRequest(query, serParams, timeoutMillis, cursorBufferSize, "", resultType, qid)
+	resp, err := s.invokeOnConnection(ctx, req, conn)
+	if err != nil {
+		return nil, err
+	}
+	metadata, page, updateCount, err := codec.DecodeSqlExecuteResponse(resp, s.serializationService)
+	if !reflect.ValueOf(err).IsNil() {
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/sql/driver/sql_service.go:129]: Due to how `DecodeSqlError` was generated (in the client binary protocol), even if the value of `err` is `nil`, it has a type which is not `nil`, so comparing it to `nil` would return true. Using `reflect.ValueOf` is an idiomatic way of checking whether a value is actually `nil`. But it seems checking it against `(*isql.Error)(nil)` is more performant.

File: sql/example_test.go (hunk 3 above)
Comment (yuce) [reply sql/example_test.go]: I've changed how default port is handled at https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/c63fe8e2987fe317b8b18296dc3adeb6043e8616

File: sql/example_test.go (hunk 2 above)
Comment (yuce) [reply sql/example_test.go]: Added at https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/c63fe8e2987fe317b8b18296dc3adeb6043e8616

File: internal/sql/driver/statement.go (hunk 1 above)
Comment (yuce) [reply internal/sql/driver/statement.go:45]: Thanks! I've added them at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/c63fe8e2987fe317b8b18296dc3adeb6043e8616

File: internal/sql/driver/sql_service.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuce) [reply internal/sql/driver/sql_service.go]: Changed at https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/c63fe8e2987fe317b8b18296dc3adeb6043e8616

File: internal/sql/driver/driver_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuce) [reply internal/sql/driver/driver_test.go]: The "properties" in the DSN are case insensitive, in order to make it more convenient, but let's discuss if you have other thoughts about it. I've fixed the DSN at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/c63fe8e2987fe317b8b18296dc3adeb6043e8616

File: internal/sql/driver/driver_test.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/hazelcast/hazelcast-go-client/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/sql/driver"
+	"github.com/hazelcast/hazelcast-go-client/types"
+)
+
+func TestParseDSN(t *testing.T) {
+	testCases := []struct {
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/sql/driver/driver_test.go:31]: Changed one of the test cases to include that at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/84ca24d8411bc0967ca5de9ee0d067d9e1ea0f22

File: internal/sql/driver/driver_it_test.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	"testing"
+
+	"github.com/hazelcast/hazelcast-go-client/internal/it"
+)
+
+func TestConnector_Driver(t *testing.T) {
+	it.SkipIf(t, "hz < 5.0")
+	db, err := sql.Open("hazelcast", "localhost:5701")
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/sql/driver/driver_it_test.go]: We already have `sql/driver/driver_it_test.go`, so removed this one.

File: client.go [hunk 10]
```diff
@@ -338,22 +260,42 @@ func (c *Client) RemoveMembershipListener(subscriptionID types.UUID) error {
...
 func (c *Client) RemoveDistributedObjectListener(ctx context.Context, subscriptionID types.UUID) error {
-	if atomic.LoadInt32(&c.state) >= stopping {
+	if c.ic.State() >= client.Stopping {
 		return hzerrors.ErrClientNotActive
 	}
 	return c.proxyManager.removeDistributedObjectEventListener(ctx, subscriptionID)
 }
 
+/*
+func (c *Client) ExecuteSQL(ctx context.Context, sql string, params ...interface{}) (*sql.ExecResult, error) {
```
[the remark this answers is not in the document]
Comment (yuce) [reply client.go]: Yes, removed at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/84ca24d8411bc0967ca5de9ee0d067d9e1ea0f22

File: client.go [hunk 11]
```diff
@@ -61,68 +44,48 @@ func StartNewClient(ctx context.Context) (*Client, error) {
 
 // StartNewClientWithConfig creates and starts a new client with the given configuration.
 func StartNewClientWithConfig(ctx context.Context, config Config) (*Client, error) {
-	if client, err := newClient(config); err != nil {
+	c, err := newClient(config)
+	if err != nil {
 		return nil, err
-	} else if err = client.start(ctx); err != nil {
+	}
+	if err = c.ic.Start(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (yuce) [reply client.go:51]: I agree. Changed at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/84ca24d8411bc0967ca5de9ee0d067d9e1ea0f22

File: examples/sql/driver/ping/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply examples/sql/driver/ping/main.go:33]: Oh sorry, I thought it was other way around. Please ignore this

File: internal/proto/codec/sql_error_codec.go [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func DecodeSqlError(frameIterator *proto.ForwardFrameIterator) *isql.Error {
+	// begin frame
+	frameIterator.Next()
+	initialFrame := frameIterator.Next()
+	code := FixSizedTypesCodec.DecodeInt(initialFrame.Content, SqlErrorCodecCodeFieldOffset)
+	originatingMemberId := FixSizedTypesCodec.DecodeUUID(initialFrame.Content, SqlErrorCodecOriginatingMemberIdFieldOffset)
+
+	message := CodecUtil.DecodeNullableForString(frameIterator)
```
Comment (emreyigit) [inline internal/proto/codec/sql_error_codec.go:49]: Second frame may contain suggestion by server. You may want to look at https://github.com/hazelcast/hazelcast-python-client/issues/470

Comment (yuce) [reply internal/proto/codec/sql_error_codec.go:49]: Thanks. This file was auto-generated. I'll check why the suggestion field was not included.

File: internal/sql/driver/rows.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+package driver
+
+import (
+	"database/sql/driver"
+	"io"
+
+	isql "github.com/hazelcast/hazelcast-go-client/internal/sql"
+)
+
+type Rows struct {
```
Comment (mdumandag) [inline internal/sql/driver/rows.go:26]: Do we use this anywhere?

File: internal/sql/driver/sql_service.go [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+	}
+	return nil
+}
+
+func (s *SQLService) executeSQL(ctx context.Context, query string, resultType byte, timeoutMillis int64, cursorBufferSize int32, params []driver.Value) (interface{}, error) {
+	serParams, err := s.serializeParams(params)
+	if err != nil {
+		return nil, err
+	}
+	conn := s.connectionManager.RandomConnection()
```
Comment (mdumandag) [inline internal/sql/driver/sql_service.go:121]: There is an extra logic in the Java side 5.0 implementation for selecting random SQL connection

File: internal/sql/driver/sql_service.go [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+	if conn == nil {
+		return nil, ihzerrors.NewIOError("no connection found", nil)
+	}
+	qid := isql.NewQueryIDFromUUID(conn.MemberUUID())
+	req := codec.EncodeSqlExecuteRequest(query, serParams, timeoutMillis, cursorBufferSize, "", resultType, qid)
```
Comment (mdumandag) [inline internal/sql/driver/sql_service.go:126]: Have you used the latest codec definitions? There is an extra boolean parameter in the latest version

Comment (mdumandag) [inline internal/sql/driver/sql_service.go:126]: Is there a reason we don't support `schema` option? I guess it is not implemented on the server yet but not sure

File: internal/sql/row.go [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+package sql
+
+type Row struct {
+	Metadata RowMetadata
+	Columns  [][]interface{}
+}
+
+func (r Row) ColumnByName(name string) interface{} {
+	panic("not implemented")
```
Comment (mdumandag) [inline internal/sql/row.go:25]: Is there a reason or is it really not implemented yet? I am wondering, if we are going to do that before merging this PR

File: serialization/api.go [hunk 17]
```diff
@@ -28,10 +35,10 @@ type IdentifiedDataSerializableFactory interface {
 // Each IdentifiedDataSerializable is created by a registered IdentifiedDataSerializableFactory.
 type IdentifiedDataSerializable interface {
 	// FactoryID returns IdentifiedDataSerializableFactory factory ID for this struct.
-	FactoryID() (factoryID int32)
+	FactoryID() int32
```
Comment (mdumandag) [inline serialization/api.go:38]: Can this be a breaking change? I mean, can there be users relying on named return values (or whatever they are called in Go)?

File: serialization/api.go [hunk 18]
```diff
@@ -155,11 +162,47 @@ type DataOutput interface {
...
 
-	// WriteBytes writes a string's characters.
+	// WriteStringBytes writes a string's characters.
 	WriteStringBytes(bytes string)
 
 	// WriteZeroBytes writes zero bytes as given length.
 	WriteZeroBytes(count int)
+
+	// WriteDate writes the given date.
+	WriteDate(t time.Time)
```
Comment (mdumandag) [inline serialization/api.go:172]: These methods do not exist in the DataOutput interface in Java

File: serialization/api.go [hunk 19]
```diff
@@ -171,158 +214,159 @@ type DataOutput interface {
...
-
 	// ReadFloat64Array returns []float64 read.
 	// It returns nil if an error is set previously.
 	ReadFloat64Array() []float64
-
 	// ReadStringArray returns []string read.
 	// It returns nil if an error is set previously.
 	ReadStringArray() []string
+	// ReadDate reads and returns a date.
+	ReadDate() time.Time
```
Comment (mdumandag) [inline serialization/api.go:277]: Same as above

File: sql/driver/driver_it_test.go [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+			SmallIntValue: 32767,
+			IntegerValue:  -27,
+			BigIntValue:   38,
+			BoolValue:     true,
+			RealValue:     -5.32,
+			DoubleValue:   12.789,
+			//DateValue:                  tt,
+			//TimeValue:                  tt,
+			//TimestampValue:             tt,
+			//TimestampWithTimeZoneValue: tt,
```
Comment (mdumandag) [inline sql/driver/driver_it_test.go:174]: Why are these commented out?

File: sql/driver/driver_it_test.go [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+	return entries, nil
+}
+
+func makeDSN(config *hz.Config) string {
+	ll := logger.InfoLevel
+	if it.TraceLoggingEnabled() {
+		ll = logger.TraceLevel
+	}
+	return fmt.Sprintf("%s;cluster.name=%s;cluster.unisocket=%t;logger.level=%s", config.Cluster.Network.Addresses[0], config.Cluster.Name, config.Cluster.Unisocket, ll)
+}
```
Comment (mdumandag) [inline sql/driver/driver_it_test.go:304]: I would like to see that we can decode every column type, to see that we are compatible with the server. Similar to https://github.com/hazelcast/hazelcast-python-client/blob/master/tests/integration/backward_compatible/sql_test.py#L631

**Review by mdumandag — COMMENTED** [review]
First iteration of the review.
I think we should definitely

- add documentation & more code samples
- improve tests

And, personally, I would like to have an API on the client as well, as I believe there might be users using the client and want to try to SQL, without creating another client

File: internal/serialization/types.go [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+
+import (
+	"math"
+	"math/big"
+	"math/bits"
+)
+
+// BigIntToJavaBytes returns a Java BigInteger compatible byte array.
+func BigIntToJavaBytes(b *big.Int) []byte {
+	i := b.BitLen()/8 + 1
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/serialization/types.go]: Thanks for catching this. I've updated big integer code and added the test above at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/b5b6bcc2e4abc9441ff18111a16085596646b5a2

File: internal/proto/codec/sql_execute_reserved_codec.go [hunk 23]
```diff
@@ -0,0 +1,71 @@
+/*
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/proto/codec/sql_execute_reserved_codec.go]: Removed at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/b5b6bcc2e4abc9441ff18111a16085596646b5a2

File: internal/proto/codec/sql_fetch_reserved_codec.go [hunk 24]
```diff
@@ -0,0 +1,62 @@
+/*
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/proto/codec/sql_fetch_reserved_codec.go]: Removed at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/b5b6bcc2e4abc9441ff18111a16085596646b5a2

File: internal/serialization/serialization.go [hunk 25]
```diff
@@ -346,6 +359,21 @@ func makeError(rec interface{}) error {
 	}
 }
 
+func dateTimeSerializer(t time.Time) pubserialization.Serializer {
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/serialization/serialization.go:362]: Updated the documentation at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/b5b6bcc2e4abc9441ff18111a16085596646b5a2

File: internal/sql/driver/driver.go [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+type Driver struct {
+}
+
+func (d *Driver) Open(name string) (driver.Conn, error) {
+	return newConn(name)
+}
+
+func ParseDSN(dsn string) (*client.Config, error) {
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/sql/driver/driver.go]: MS SQL driver uses semi-columns to separate options, that's where I got the inspiration, but most other drivers use the URL format. I've converted to that scheme at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/7f89f909d650e36ced8c46d5434858c3d34761cc

File: internal/sql/driver/result.go [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+			return io.EOF
+		}
+		if err := r.fetchNextPage(); err != nil {
+			return err
+		}
+		// after fetching next page, the page and its cols change, so have to refresh them
+		cols = r.page.Columns
+	}
+	for i := 0; i < len(cols); i++ {
+		dest[i] = cols[i][r.index]
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/sql/driver/result.go:96]: The query result is not concurrency-safe (similar to other drivers). I should've documented it, but will do that if I already  didn't.

File: internal/sql/driver/rows.go (hunk 13 above)
Comment (yuce) [reply internal/sql/driver/rows.go]: No, removed at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/ccf3a8bf8ed36579f2705198ef71f58d0d17ec53

File: internal/sql/row.go (hunk 16 above)
Comment (yuce) [reply internal/sql/row.go]: Removed the unused `Row` struct at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/ccf3a8bf8ed36579f2705198ef71f58d0d17ec53

File: sql/driver/driver_it_test.go (hunk 20 above)
Comment (yuce) [reply sql/driver/driver_it_test.go]: `assert.Equal` doesn't work with them. I'll add date/time specific tests.

File: serialization/api.go (hunk 17 above)
Comment (yuce) [reply serialization/api.go:38]: No, the name of the return variable has no meaning to the caller, and the caller cannot access that.

File: internal/sql/driver/sql_service.go (hunk 15 above)
Comment (yuce) [reply internal/sql/driver/sql_service.go]: Updated the codecs at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/32115550b720632d1bb90fa96667aaf88ebf887f

File: internal/proto/codec/builtin.go [hunk 28]
```diff
@@ -412,24 +444,47 @@ func (fixSizedTypesCodec) DecodeLong(buffer []byte, offset int32) int64 {
...
 }
 
-func (fixSizedTypesCodec) EncodeBoolean(buffer []byte, offset int32, value bool) {
-	if value {
-		buffer[offset] = 1
-	} else {
-		buffer[offset] = 0
-	}
+func (fixSizedTypesCodec) EncodeFloat(buffer []byte, offset int32, value float32) {
+	binary.LittleEndian.PutUint32(buffer[offset:], math.Float32bits(value))
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/proto/codec/builtin.go:448]: `binary.LittleEndian` type has methods for only unsigned integers. The unsigned signed integers have the same bits as long as their sizes match.

File: internal/proto/codec/builtin.go [hunk 29]
```diff
@@ -396,6 +400,34 @@ type fixSizedTypesCodec struct{}
...
+func (fixSizedTypesCodec) EncodeByte(buffer []byte, offset int32, value byte) {
+	buffer[offset] = value
+}
+
+func (fixSizedTypesCodec) DecodeByte(buffer []byte, offset int32) byte {
+	return buffer[offset]
+}
+
+func (fixSizedTypesCodec) EncodeShort(buffer []byte, offset, value int32) {
+	binary.LittleEndian.PutUint16(buffer[offset:], uint16(value))
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/proto/codec/builtin.go:424]: binary.LittleEndian type has methods for only unsigned integers. The unsigned signed integers have the same bits as long as their sizes match.

File: internal/proto/codec/builtin.go [hunk 30]
```diff
@@ -412,24 +444,47 @@ func (fixSizedTypesCodec) DecodeLong(buffer []byte, offset int32) int64 {
...
-func (fixSizedTypesCodec) EncodeByte(buffer []byte, offset int32, value byte) {
-	buffer[offset] = value
+func (fixSizedTypesCodec) EncodeDouble(buffer []byte, offset int32, value float64) {
+	binary.LittleEndian.PutUint64(buffer[offset:], math.Float64bits(value))
 }
 
-func (fixSizedTypesCodec) DecodeByte(buffer []byte, offset int32) byte {
-	return buffer[offset]
+func (fixSizedTypesCodec) DecodeDouble(buffer []byte, offset int32) float64 {
+	return math.Float64frombits(binary.LittleEndian.Uint64(buffer[offset:]))
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/proto/codec/builtin.go:460]: No, a slice contains just the offset and the length. In this case specifying the end index doesn't matter.

File: internal/proto/codec/sql_error_codec.go (hunk 12 above)
Comment (yuce) [reply internal/proto/codec/sql_error_codec.go:50]: Updated the codecs at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/32115550b720632d1bb90fa96667aaf88ebf887f

File: internal/serialization/object_data.go [hunk 31]
```diff
@@ -271,6 +274,86 @@ func (o *ObjectDataOutput) WriteStringBytes(v string) {
 	o.writeStringBytes([]rune(v))
 }
 
+func (o *ObjectDataOutput) WriteRawBytes(b []byte) {
+	o.EnsureAvailable(ByteSizeInBytes * len(b))
+	o.position += int32(copy(o.buffer[o.position:], b))
+}
+
+func (o *ObjectDataOutput) WriteDate(t time.Time) {
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/serialization/object_data.go]: Removed them at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/3f2603613904c066ed58d86749debf6358fe5cec

File: internal/sql/driver/result.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (yuce) [reply internal/sql/driver/result.go:107]: Removed those atomics at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/cb3d75f2cf3c0056e101535c8c55a5f9acd75ce9

File: internal/sql/driver/sql_service.go (hunk 14 above)
Comment (yuce) [reply internal/sql/driver/sql_service.go]: updated at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/c079d377f9fc7898afdb12cd40a624e78c4b6c8f

File: serialization/api.go (hunk 18 above)
Comment (yuce) [reply serialization/api.go]: Removed them at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/3f2603613904c066ed58d86749debf6358fe5cec

File: serialization/api.go (hunk 19 above)
Comment (yuce) [reply serialization/api.go]: Removed at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/3f2603613904c066ed58d86749debf6358fe5cec

**Review by utku-caglayan — APPROVED** [review]

File: internal/serialization/portable_serializer_test.go [hunk 32]
```diff
@@ -257,6 +262,8 @@ func (f *fake) WritePortable(writer serialization.PortableWriter) {
 	writer.WriteFloat64Array("f64Arr", f.f64Arr)
 	writer.WriteStringArray("utfArr", f.utfArr)
 	writer.WritePortableArray("portableArr", f.portableArr)
+	writer.WriteDate("nilDate", f.nilDate)
+	writer.WriteDate("date", &f.date)
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/serialization/portable_serializer_test.go:278]: Added more at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/db347f022df35b809058a714e6bb42db3964b9d2 and https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/c1a220ffd09b7eb9d3144c3f6c37dd90aaac95e4

File: sql/driver/driver_it_test.go (hunk 21 above)
Comment (yuce) [reply sql/driver/driver_it_test.go:512]: Adde more tests at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/db347f022df35b809058a714e6bb42db3964b9d2 All column types should be tested now.

File: internal/sql/driver/sql_service.go (hunk 15 above)
Comment (yuce) [reply internal/sql/driver/sql_service.go]: That was my impression as well. I did not include it since it is not supported yet on the server.

File: sql/driver/driver_it_test.go (hunk 20 above)
Comment (yuce) [reply sql/driver/driver_it_test.go]: They are at: https://github.com/yuce/hazelcast-go-client/blob/f40e9a590439a52268f01e2f6d96111202b3a10c/sql/driver/driver_it_test.go#L312

**Review by emreyigit — APPROVED** [review]

File: internal/cluster/cluster_service.go [hunk 33]
```diff
@@ -284,11 +312,54 @@ func (m *membersMap) reset() {
...
+			vs[next] = v
+			vv[next] = mem.Version
+			count[next]++
+			next++
+		} else {
+			return 0, fmt.Errorf("more than 2 distinct member versions found: %s, %s, %s", vv[0], vv[1], mem.Version)
+		}
+		if i > majority {
+			break
+		}
```
Comment (mdumandag) [inline internal/cluster/cluster_service.go:354]: Is this logic correct? I mean, we should break only if one of the groups is larger than the majority.

File: internal/cluster/cluster_service.go [hunk 34]
```diff
@@ -284,11 +312,54 @@ func (m *membersMap) reset() {
...
+	}
+	if count[0] == 0 {
+		// there are no data members
+		return 0, nil
+	}
+	if count[0] >= count[1] {
```
Comment (mdumandag) [inline internal/cluster/cluster_service.go:360]: There is an extra logic in the Java side, once the counts are equal, they return the newer group

File: internal/cluster/connection_manager.go [hunk 35]
```diff
@@ -306,6 +306,35 @@ func (m *ConnectionManager) RandomConnection() *Connection {
 	return m.connMap.RandomConn()
 }
 
+func (m *ConnectionManager) SQLConnection() *Connection {
+	if m.smartRouting {
+		if member := m.clusterService.SQLMember(); member != nil {
```
Comment (mdumandag) [inline internal/cluster/connection_manager.go:311]: There is a loop in the Java side (SQL_CONNECTION_RANDOM_ATTEMPTS times) with an explanation on top of it. Is this required here?

File: internal/serialization/object_data.go [hunk 36]
```diff
@@ -739,3 +743,149 @@ func (e *EmptyObjectDataOutput) WriteStringArray([]string) {}
 func (e *EmptyObjectDataOutput) WriteStringBytes(string) {}
 
 func (e *EmptyObjectDataOutput) WriteZeroBytes(int) {}
+
+func WriteDate(o serialization.DataOutput, t time.Time) {
```
Comment (mdumandag) [inline internal/serialization/object_data.go:747]: Are these methods leftovers?

File: internal/serialization/big_int.go [hunk 37]
```diff
@@ -0,0 +1,293 @@
+/*
```
Comment (mdumandag) [inline internal/serialization/big_int.go:1]: Honestly, the code is too complex to review but I will assume it is correct, given that we do have lots of tests covering edge cases, small and large value :+1:

File: examples/sql/driver/ping/main.go [hunk 38]
```diff
@@ -0,0 +1,39 @@
...
+func main() {
+	log.Println("Creating the database value.")
+	db, err := sql.Open("hazelcast", "hz://?logger.level=info")
+	if err != nil {
+		panic(err)
+	}
+	defer db.Close()
+	log.Println("The database value was created. Note that the connection to Hazelcast is not established at this point.")
+	log.Println("Sending a ping request to establish the connection...")
+	if err := db.Ping(); err != nil {
```
[the remark this answers is not in the document]
Comment (yuce) [reply examples/sql/driver/ping/main.go:35]: The underlying Hazelcast client is lazy created. Creating the "sql.Db" instance using `sql.Open` doesn't actually create the Hazelcast client. It is only created when necessary. `db.Ping` makes sure the client is connected, without actually sending a request. database/sql package handles that stuff automatically.

File: internal/cluster/connection_manager.go (hunk 35 above)
Comment (yuce) [reply internal/cluster/connection_manager.go:311]: We talked with @sancar about that. `SQL_CONNECTION_RANDOM_ATTEMPTS` doesn't make much sense. We should probably change other clients to remove that.

File: internal/serialization/object_data.go (hunk 36 above)
Comment (yuce) [reply internal/serialization/object_data.go:747]: No, those functions are used by `JavaLocalDateSerializer`, etc.

File: internal/serialization/big_int.go (hunk 37 above)
Comment (yuce) [reply internal/serialization/big_int.go:1]: :pray:

File: internal/cluster/cluster_service.go (hunk 33 above)
Comment (yuce) [reply internal/cluster/cluster_service.go:354]: `i` corresponds to `currentCount` in the Java side, so I think that code may be also wrong. And in the Java side that piece of code only runs when there is the `localMember` (i.e., never) so I'll just remove it.

Comment (yuce) [reply internal/cluster/cluster_service.go]: Removed at: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/67c7887a6db6650f61b79de879d8b6456be75afe

File: internal/cluster/cluster_service.go (hunk 34 above)
Comment (yuce) [reply internal/cluster/cluster_service.go]: Yeah, we should probably simplify all that logic, but I've added that for the time being: https://github.com/hazelcast/hazelcast-go-client/pull/695/commits/67c7887a6db6650f61b79de879d8b6456be75afe

**Review by mdumandag — APPROVED** [review]

**Comment by yuce** [comment]
@mdumandag @utku-caglayan @emreyigit Thanks for the review!
